File: week_1/code/balanced_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        """
        Given a binary tree, determine if it is height-balanced.
        For this problem, a height-balanced binary tree is defined as:
        a binary tree in which the left and right subtrees of every node differ in height by no more than 1.
        """
        def dfs(node):
            if not node:
                return 0
            
            left = dfs(node.left)
            if left == -1: return -1
            
            right = dfs(node.right)
            if right == -1: return -1
            
            if abs(left - right) > 1:
                return -1
                
            return 1 + max(left, right)
            
        return dfs(root) != -1

# Helper to build tree from level order list
def buildTree(arr):
    if not arr: return None
    root = TreeNode(arr[0])
    queue = [root]
    i = 1
    while queue and i < len(arr):
        node = queue.pop(0)
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    return root

# Test cases
if __name__ == "__main__":
    sol = Solution()
    
    # Test 1: [3,9,20,null,null,15,7] (Balanced)
    root1 = buildTree([3,9,20,None,None,15,7])
    print(f"Test 1: {sol.isBalanced(root1)} (Expected: True)")

    # Test 2: [1,2,2,3,3,null,null,4,4] (Unbalanced)
    root2 = buildTree([1,2,2,3,3,None,None,4,4])
    print(f"Test 2: {sol.isBalanced(root2)} (Expected: False)")

File: week_1/code/balanced_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
    function dfs(node) {
        if (!node) return 0;

        let left = dfs(node.left);
        if (left === -1) return -1;

        let right = dfs(node.right);
        if (right === -1) return -1;

        if (Math.abs(left - right) > 1) return -1;

        return 1 + Math.max(left, right);
    }

    return dfs(root) !== -1;
};

// Helper to build tree from array
function buildTree(arr) {
    if (!arr.length) return null;
    let root = new TreeNode(arr[0]);
    let queue = [root];
    let i = 1;
    while (queue.length > 0 && i < arr.length) {
        let node = queue.shift();
        if (i < arr.length && arr[i] !== null) {
            node.left = new TreeNode(arr[i]);
            queue.push(node.left);
        }
        i++;
        if (i < arr.length && arr[i] !== null) {
            node.right = new TreeNode(arr[i]);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

// Test cases
// Test 1: [3,9,20,null,null,15,7] (Balanced)
let root1 = buildTree([3, 9, 20, null, null, 15, 7]);
console.log("Test 1:", isBalanced(root1), "(Expected: true)");

// Test 2: [1,2,2,3,3,null,null,4,4] (Unbalanced)
let root2 = buildTree([1, 2, 2, 3, 3, null, null, 4, 4]);
console.log("Test 2:", isBalanced(root2), "(Expected: false)");
